arch/xtensa/src/esp32s3/esp32s3_tickless.c

```diff
@@ -0,0 +1,484 @@
+ * specific interval timer implementation:
+ *
+ *   void sched_timer_expiration(void):  Called by the platform-specific
+ *     logic when the interval timer expires.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <time.h>
+#include <assert.h>
+
+#include <nuttx/irq.h>
+#include <nuttx/arch.h>
+#include <nuttx/clock.h>
+#include <arch/board/board.h>
+#include <arch/irq.h>
+
+#include "xtensa.h"
+#include "chip.h"
+#include "esp32s3_irq.h"
+#include "hardware/esp32s3_systimer.h"
+#include "hardware/esp32s3_system.h"
+#include "hardware/esp32s3_soc.h"
+
+#ifdef CONFIG_SCHED_TICKLESS
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define ESP32S3_SYSTIMER_TICKS_PER_SEC  (16 * 1000 * 1000)
+
+#define CTICK_PER_SEC         (ESP32S3_SYSTIMER_TICKS_PER_SEC)
+#define CTICK_PER_USEC        (CTICK_PER_SEC / USEC_PER_SEC)
+
+#define SEC_2_CTICK(s)        ((s) * CTICK_PER_SEC)
+#define USEC_2_CTICK(us)      ((us) * CTICK_PER_USEC)
+#define NSEC_2_CTICK(nsec)    (((nsec) * CTICK_PER_USEC) / NSEC_PER_USEC)
+
+#define CTICK_2_SEC(tick)     ((tick) / CTICK_PER_SEC)
+#define CTICK_2_USEC(tick)    ((tick) / CTICK_PER_USEC)
+#define CTICK_2_NSEC(tick)    ((tick) * 1000 / CTICK_PER_USEC)
+
+#define CPU_TICKS_MAX         (UINT32_MAX / 4 * 3)
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static inline uint64_t up_tmr_getcounter(void);
+static inline uint64_t up_tmr_getalarmvalue(void);
+static void IRAM_ATTR up_tmr_setcounter(uint64_t ticks);
+static int IRAM_ATTR up_timer_expire(int irq, void *context, void *arg);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static bool g_timer_started; /* Whether an interval timer is being started */
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_tmr_getcounter
+ *
+ * Description:
+ *   Return the total ticks of system since power-on.
+ *
+ * Input Parameters:
+ *   None.
+ *
+ * Returned Value:
+ *   Total system ticks.
+ *
+ ****************************************************************************/
+
+static inline uint64_t up_tmr_getcounter(void)
```

Comment:
why do you use `up_` notation for local functions?

Reply:
Interesting question. This driver is a port from another ESP32 chip, the `up_` prefix didn't caught my attention before.
It seems pointless for these internal functions to be prefixed with `up_`. I will rename them.
Thanks.

Reply:
Thank you

Reply:
Done, renamed the private `up_` prefixed functions to `tickless_`.